Various TypeScript related resolutions, fixes, additions, and electric boogaloo

favna opened the pull request:
Pretty much thoroughly discussed on Discord already. Opening up the PR so it can be merged.

- Fix the lib to actually compile
- Introduce Jest as test framework to test all sites
  - It also tests all other exported functions from index.ts
- Introduce `terser-folder` to compress and mangle the build
- Always run lint and test prior to building
- Change to MIT license to resolve #22 
- Remove youhate.us since it's dead
- Address some feedback you already gave on Discord prior to PR

favna committed:
TypeScript fixes and resolutions

Now it compiles and runs properly, yay!

favna committed:
Code formatting

favna committed:
Fixed for derpibooru

favna committed:
Change to MIT license

Resolves #22

favna committed:
Fix site resolution

favna committed:
Write tests

favna committed:
Cleanup package

Commit:
Remove youhate.us

Commit:
Add package-lock

Commit:
Fix up nsfw prop

Signed-off-by: Jeroen Claassens <support@fanva.xyz>

Commit:
Update example usage in readme

Commit:
Update readme contributors

Commit:
Update example

Commit:
Forgot to remove common-tags type def

Commit:
Run tests prebuild

Commit:
Update examples

favna committed:
Add missing semi's and revert to function keyword

AtoraSuunva commented on src/Constants.ts:
```diff
@@ -1,99 +1,104 @@
+  search: (tags: string | string[], SearchArgs: SearchParameters) => any;
+};
+
+export interface SMap {
```
I would probably be a better idea to specify this as a generic with `SMap<V>` rather than unison `SiteInfo & searchFunction`

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/Constants.ts:
```diff
@@ -1,99 +1,104 @@
+ * The user-agent to use for searches
+ * @private
+ */
+export const userAgent: string = `booru (https://github.com/AtlasTheBot/booru)`;
```
I would prefer keeping the booru version in the userAgent (in case it's ever needed to debug which version is causing issues at some point)

favna replied:
Messaged about this on Discord

favna replied:
Changing would have the following impact:

```diff
diff --git a/package.json b/package.json
index fdd4f72..90d7dc4 100644
--- a/package.json
+++ b/package.json
@@ -4,8 +4,8 @@
   "description": "Search (and do other things) on a bunch of different boorus!",
   "author": "AtlasTheBot",
   "license": "MIT",
-  "main": "./dist/index.js",
-  "typings": "./dist/index.d.ts",
+  "main": "./dist/src/index.js",
+  "typings": "./dist/src/index.d.ts",
   "runkitExampleFilename": "./dist/example.js",
   "homepage": "https://github.com/AtlasTheBot/booru#readme",
   "repository": {
```

```diff
diff --git a/src/Constants.ts b/src/Constants.ts
index ad97fe3..59212c3 100644
--- a/src/Constants.ts
+++ b/src/Constants.ts
@@ -1,4 +1,5 @@
 import { RequestInit } from 'node-fetch';
+import pkg from '../package.json';
 import siteJson from './sites.json';
 import SearchParameters from './structures/SearchParameters.js';
 import Site from './structures/Site';
@@ -54,7 +55,7 @@ export class BooruError extends Error {
  * The user-agent to use for searches
  * @private
  */
-export const userAgent: string = `booru (https://github.com/AtlasTheBot/booru)`;
+export const userAgent: string = `booru@${pkg.version} (https://github.com/AtlasTheBot/booru)`;

 /**
  * Expands tags based on a simple map, used for gelbooru/safebooru/etc compat :(
```

```diff
diff --git a/tsconfig.json b/tsconfig.json
index d9f5975..a3204f0 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "outDir": "./dist/",
-    "rootDir": "./src",
+    "rootDir": "./",
     "forceConsistentCasingInFileNames": true,
     "lib": ["esnext"],
     "target": "esnext",
```

```
$ npm publish --dry-run
npm notice
npm notice package: booru@2.0.0
npm notice === Tarball Contents ===
npm notice 1.8kB package.json
npm notice 1.1kB LICENSE.md
npm notice 4.3kB readme.md
npm notice 1.8kB dist/example.js
npm notice 2.0kB dist/package.json
npm notice 814B  dist/src/boorus/Booru.d.ts
npm notice 2.0kB dist/src/boorus/Booru.js
npm notice 394B  dist/src/boorus/Derpibooru.d.ts
npm notice 779B  dist/src/boorus/Derpibooru.js
npm notice 167B  dist/src/boorus/XmlBooru.d.ts
npm notice 309B  dist/src/boorus/XmlBooru.js
npm notice 681B  dist/src/Constants.d.ts
npm notice 1.2kB dist/src/Constants.js
npm notice 653B  dist/src/index.d.ts
npm notice 2.0kB dist/src/index.js
npm notice 5.8kB dist/src/sites.json
npm notice 212B  dist/src/structures/InternalSearchParameters.d.ts
npm notice 68B   dist/src/structures/InternalSearchParameters.js
npm notice 889B  dist/src/structures/Post.d.ts
npm notice 2.4kB dist/src/structures/Post.js
npm notice 133B  dist/src/structures/SearchParameters.d.ts
npm notice 68B   dist/src/structures/SearchParameters.js
npm notice 623B  dist/src/structures/SearchResults.d.ts
npm notice 924B  dist/src/structures/SearchResults.js
npm notice 332B  dist/src/structures/Site.d.ts
npm notice 350B  dist/src/structures/Site.js
npm notice 84B   dist/src/structures/SiteApi.d.ts
npm notice 68B   dist/src/structures/SiteApi.js
npm notice 277B  dist/src/structures/SiteInfo.d.ts
npm notice 68B   dist/src/structures/SiteInfo.js
npm notice 462B  dist/src/Utils.d.ts
npm notice 1.3kB dist/src/Utils.js
npm notice === Tarball Details ===
npm notice name:          booru
npm notice version:       2.0.0
npm notice package size:  9.8 kB
npm notice unpacked size: 34.0 kB
npm notice shasum:        a8e2e461a182ebc211e95d35ba178e8228edbdad
npm notice integrity:     sha512-WkXmOL4jRP7C3[...]23RVCfweIwndQ==
npm notice total files:   32
npm notice
+ booru@2.0.0
```

![image](https://user-images.githubusercontent.com/4019718/53326530-1d7ee600-38e6-11e9-9616-12978e789141.png)

favna replied:
***personally*** I'd say don't and if someone creates an issue on the project they are just obliged to list the version. For example I could create an `ISSUE_TEMPLATE.md` file which is those templates you see in other projects - and list that if people do not fill out the template they'll first get asked to do so before they get support.

AtoraSuunva replied:
Yeah, I'd prefer not having the version listed over having an odd directory structure

AtoraSuunva commented on example.js:
```diff
@@ -1,45 +1,39 @@
-  // In the future, things like `e9.favorite(Post)` or `e9.fetchComments(Post)`
-  // Will be available (and <Post>.favorite())
-}
+const Booru = require('booru');
```
Personally prefer no semicolon style, although for an example I don't mind that much

favna replied:
So... what do? Remove it or keep it? I don't really care either way.

AtoraSuunva replied:
Preferably remove them, to keep it consistent with the older style

favna replied:
Resolved in 6104080861912d66f9f3b1d47ffbd5738ef4941f

AtoraSuunva commented on package.json:
```diff
@@ -1,46 +1,74 @@
+  "version": "2.0.0",
+  "description": "Search (and do other things) on a bunch of different boorus!",
+  "author": "AtlasTheBot",
+  "license": "GPL-3.0",
```
License should be MIT

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on package.json:
```diff
@@ -1,46 +1,74 @@
+    "test": "jest --no-cache",
+    "lint": "tslint --fix -p . -c ./tslint.json ./src/**/*.ts",
+    "clean": "rimraf ./dist",
+    "prebuild": "yarn clean && yarn lint && yarn test",
```
Might be better to `npm` instead? Personally don't have yarn installed so this just errors out

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on package.json:
```diff
@@ -1,46 +1,74 @@
+    "derpibooru"
+  ],
+  "engines": {
+    "node": ">=8.0.0"
```
Should bump this up to new LTS (10.15.1) so we can target esnext instead and avoid the overhead of es6 compat

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9 - Also see discussion about esnext in feedback below

AtoraSuunva commented on readme.md:
```diff
@@ -1,136 +1,134 @@
+  return console.log(posts[0].fileUrl);
+};
+
+console.log(Booru.SiteMap); // you can also check the sites and the options for each
```
Would prefer to keep it as `booru.sites`, rather than introduce another breaking change for no good reason

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on readme.md:
```diff
@@ -1,136 +1,134 @@
+
+### License?
+
+[It's GPLv3](http://choosealicense.com/licenses/gpl-3.0/)
```
GPLv3 => MIT here as well, although this is more a note for myself lol

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/Constants.ts:
```diff
@@ -1,99 +1,104 @@
+import Site from './structures/Site';
+import SiteInfo from './structures/SiteInfo';
+
+type searchFunction = {
```
What's the use of this type? I can only find it in this file here and on line 11 (SMap)

favna replied:
Addressed in Discord and also addressed in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/Constants.ts:
```diff
@@ -1,99 +1,104 @@
+/**
+ * A map of site url/{@link SiteInfo}
+ */
+export const SiteMap: SMap = sites as any;
```
As before:

> Would prefer to keep it as `booru.sites`, rather than introduce another breaking change for no good reason

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/Constants.ts:
```diff
@@ -1,99 +1,104 @@
+
+/**
+ * The default options to use for requests
+ * <p>I could document this better but meh
```
Now that it's typed this line could be removed

favna replied:
What line exactly do you mean? I think you linked the wrong lines because I just see some JSDoc here?

AtoraSuunva commented on src/Utils.ts:
```diff
@@ -1,145 +1,133 @@
+ * @param  {String} domain The site to resolveSite
+ * @return {String?} null if site is not supported, the site otherwise
+ */
+export const resolveSite = (domain: string): string | null => {
```
Would prefer `export function` over arrow notation but that's more nitpicking

favna replied:
I missed this one and some others too... Resolved them all in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/Utils.ts:
```diff
@@ -1,145 +1,133 @@
+  domain = domain.toLowerCase();
+
+  for (const site in sites) {
+    if (site === domain || (sites[site].aliases as string[]).includes(domain)) {
```
`as string[]` cast isn't needed if SMap is generified and you use `SMap<SiteInfo>` as the type for sites iirc

favna replied:
It is, generic typing is not involved here. Without casting it to a `string[]` it throws this: 
![image](https://user-images.githubusercontent.com/4019718/53306910-4c1fa100-3893-11e9-98bd-a90c3c1aa54b.png)

AtoraSuunva replied:
Hm, in my version I used `SMap<SiteInfo>` as the type for `sites` and `tsc` didn't error

favna replied:
Same. It worked because you had `SMap<any>` for your `booruCache` in `index.ts`.

Previously your SiteInfo had this:

https://github.com/AtlasTheBot/booru/blob/7f3c9ea169551920631be1bd1c59f69bf64f585b/src/structures/SiteInfo.ts#L13-L16

Which when specifying the type of `booruCache` I changed to:

```ts
 type?: string 
 /** The aliases of this site */ 
 aliases?: string[] 
 /** If this site serves NSFW posts or not */
```

Otherwise this was thrown:

![image](https://user-images.githubusercontent.com/4019718/53325621-03440880-38e4-11e9-8af5-b4e0f1c93dec.png)

So there were 3 solutions:

1. Mark `aliases` as optional in `SiteInfo` and cast it to `string[]` later while keeping the typings for booruCache
2. Remove typing for booruCache, keep as is
3. Hackily include all those parameters in the Booru class

I went for option 1.

favna replied:
Changing that it is not required would have the following impact:

```diff
diff --git a/src/Constants.ts b/src/Constants.ts
index ad97fe3..2258b79 100644
--- a/src/Constants.ts
+++ b/src/Constants.ts
@@ -1,13 +1,8 @@
 import { RequestInit } from 'node-fetch';
 import siteJson from './sites.json';
-import SearchParameters from './structures/SearchParameters.js';
 import Site from './structures/Site';
 import SiteInfo from './structures/SiteInfo';

-export type searchFunction = {
-  search: (tags: string | string[], SearchArgs: SearchParameters) => any;
-};
-
 export interface SMap<V> {
   [key: string]: V;
 }
```

```diff
diff --git a/src/Utils.ts b/src/Utils.ts
index e7e57a3..e93818d 100644
--- a/src/Utils.ts
+++ b/src/Utils.ts
@@ -23,7 +23,7 @@ export function resolveSite (domain: string): string | null {
   domain = domain.toLowerCase();

   for (const site in sites) {
-    if (site === domain || (sites[site].aliases as string[]).includes(domain)) {
+    if (site === domain || sites[site].aliases.includes(domain)) {
       return site;
     }
   }
```

```diff
diff --git a/src/index.ts b/src/index.ts
index 70e929c..bb3f3bd 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -2,12 +2,11 @@ import { deprecate } from 'util';
 import Booru from './boorus/Booru';
 import Derpibooru from './boorus/Derpibooru';
 import XmlBooru from './boorus/XmlBooru';
-import { BooruError, searchFunction, sites, SMap } from './Constants';
+import { BooruError, sites, SMap } from './Constants';
 import Post from './structures/Post';
 import SearchParameters from './structures/SearchParameters';
 import SearchResults from './structures/SearchResults';
 import Site from './structures/Site';
-import SiteInfo from './structures/SiteInfo';
 import { resolveSite } from './Utils';

 const BooruTypes: any = {
@@ -15,7 +14,7 @@ const BooruTypes: any = {
   xml: XmlBooru,
 };

-const booruCache: SMap<SiteInfo & searchFunction> = {};
+const booruCache: SMap<any> = {};

 /**
  * Create a new booru, if special type, use that booru, else use default Booru
```

```diff
diff --git a/src/structures/SiteInfo.ts b/src/structures/SiteInfo.ts
index c35dcf5..7ff86b3 100644
--- a/src/structures/SiteInfo.ts
+++ b/src/structures/SiteInfo.ts
@@ -9,17 +9,17 @@ export default interface SiteInfo {
   /** The domain of the Site (the "google.com" part of "https://google.com/foo") */
   domain: string;
   /** The type of this site (json/xml/derpi) */
-  type?: string;
+  type: string;
   /** The aliases of this site */
-  aliases?: string[];
+  aliases: string[];
   /** If this site serves NSFW posts or not */
-  nsfw?: boolean;
+  nsfw: boolean;
   /** An object representing the api of this site */
-  api?: SiteApi;
+  api: SiteApi;
   /** The url query param to paginate on the site */
   paginate?: string;
   /** If the site supports `order:random` */
-  random?: boolean | string;
+  random: boolean | string;
   /** The url query param for tags */
   tagQuery?: string;
   /** If this site supports only http:// */
```

AtoraSuunva replied:
I'll look more into detail about this later, if I can't figure something else out I'll stick with option 1

AtoraSuunva replied:
booruCache is fine if you declare it as `SMap<Booru>`, at least for me ![booruCache](https://user-images.githubusercontent.com/15206032/53346052-2388bd00-390e-11e9-8af9-63489ef3afcb.png)

favna replied:
Well that should finally be the last one then - addressed this in 7b6490a218bc6a9aded6fe2f70673334ddd7ccb1

AtoraSuunva commented on src/boorus/Booru.ts:
```diff
@@ -1,180 +1,181 @@
+      return xml ? await jsonfy(response as unknown as string) : response;
+    } catch (err) {
+      if ((err as FetchError).type === 'invalid-json') return '';
+      return err;
```
Should be `throw err`, since returning it just resolves the promise with an Error rather than reject it

favna replied:
Resolved in 6f5f617cc7c6f818a81c69fa4f5d5ab6c2ec31c9

AtoraSuunva commented on src/index.ts:
```diff
@@ -1,121 +1,98 @@
+ * @return {Promise<Post[]>} Array of {@link Post} objects
+ */
+export async function commonfy (images: Post[]): Promise<any> {
+  return deprecate(() => images, 'Common is now deprecated, just access the properties directly');
```
This will issue a deprecation warning on every use of commonfy, rather than just once per run

favna replied:
I'm not exactly sure how I can fix this?

AtoraSuunva replied:
Could just leave it as how I had it before

```ts
const _commonfy =
  deprecate(() => { }, 'Common is now deprecated, just access the properties directly')

/**
 * Deprecated, now a noop
 * <p>This will be removed *soon* please stop using it</p>
 * <p>Just access <code>&lt;{@link Post}&gt;.prop</code>, no need to commonfy anymore
 *
 * @deprecated Just use <code>&lt;{@link Post}&gt;.prop</code> instead
 * @param  {Post[]} images   Array of {@link Post} objects
 * @return {Promise<Post[]>} Array of {@link Post} objects
 */
export function commonfy(images: Post[]): Promise<Post[]> {
  _commonfy()
  return Promise.resolve(images)
}
```

favna replied:
Reverted as suggested aside from some naming conventions (camelCase function for example) so resolved in 6104080861912d66f9f3b1d47ffbd5738ef4941f

AtoraSuunva commented on tsconfig.json:
```diff
@@ -1,25 +1,25 @@
+    "rootDir": "./src",
+    "forceConsistentCasingInFileNames": true,
+    "lib": ["es2018"],
     "target": "es6",
```
Potentially target `esnext` instead? Would reduce overhead of polyfill (and output less ugly code lol). Mostly since I hope everyone is on current LTS :(

favna replied:
I held them side-by-side and the primary difference is that when targeting `esnext` it doesn't need a polyfill for `await`. However it should be noted that postbuild `uglifyjs-folder` is ran over the code completely compressing and mangling it so at the end of the day it's 1 line minified to the extreme already. To give an idea, this is `index.js` postbuild (when targeting `es6` that is):

![image](https://user-images.githubusercontent.com/4019718/53307414-96efe780-3898-11e9-8c17-a464cedbb8f4.png)


Don't get me wrong - I do not necessarily disagree... but the gain is very minimal.

favna replied:
Oh and this would be `index.js` when targeting `esnext`:

![image](https://user-images.githubusercontent.com/4019718/53307403-745dce80-3898-11e9-82fb-ae0383868a61.png)

favna replied:
Update: now it uses `terser-folder`. Updated the screenshots above accordingly

Terser is more up2date than uglifyjs as described [in their description]([npmjs.com/package/terser#why-choose-terser](https://www.npmjs.com/package/terser#why-choose-terser))

Also it is future proofing, especially needed if we're going to target esnext.

Also Also it will be default in webpack starting v5 so like.. community says it all right?

AtoraSuunva replied:
Hm, the difference might be minimal, although when targeting LTS 10 might as well target esnext as it fully supports it

webpack might be a problem, if someone using it is trying to target older browsers without async/await support, although I'm not sure if anyone is/will ever webpack booru

favna replied:
Doubt it and if they do they should provide the necessary polyfills themselve through something like coreJS IMHO.

Changed to ESNext so resolved in 6104080861912d66f9f3b1d47ffbd5738ef4941f

AtoraSuunva reviewed (changes requested):
Check the comments I've left, but more generally I prefer no semicolon style over semicolons

favna committed:
Address feedback

- Revert to SMap<V> Generic
- Use NPM in package scripts and CircleCI
- Revert to "sites" as exported member, instead of SiteMap
- Properly reference MIT license
- Target NodeJS LTS (v10) at time of writing
- Further revert to function keyword
- Throw errors, not return them

Commit:
Use terser instead of uglifyjs

Terser is more up2date than uglifyjs as described [in their
description](https://www.npmjs.com/package/terser#why-choose-terser)

Also it is future proofing, especially needed if we're going to target
esnext.

Also Also it will be default in webpack starting v5 so like.. community
speaks right?

Commit:
Address feedback part 2

- Change to lib and target esnext
- Include DS_Store in gitignore
- Revert commonfy back to previous state
- Make example runnable and remove semi's as requested
- Remove requirement of casting to string[] for sites


Signed-off-by: Jeroen Claassens <support@fanva.xyz>

Commit:
Address feedback concerning casting and typing of booruCache

Signed-off-by: Jeroen Claassens <support@fanva.xyz>
